### PR TITLE
Implement no-async-describe rule

### PR DESCRIPTION
**Pull request by papandreou**
I've seen `describe('my test', async function () { ... })` cause subtle problems, so I think it makes sense to add a rule for it.

### CODE REVIEW

File: lib/rules/no-async-describe.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    function isFunction(node) {
+        return (
+            node.type === 'FunctionExpression' ||
+            node.type === 'FunctionDeclaration' ||
+            node.type === 'ArrowFunctionExpression'
+        );
+    }
+
+    function containsDirectAwait(node) {
```
Comment (papandreou) [inline lib/rules/no-async-describe.js]: I don't know if this could be implemented with a funky ramda function that looks for a nested object with a `type` of `AwaitExpression` not contained within one with a `type` of `FunctionExpression` etc. -- but I have to admit that I'm not really familiar with that.

File: docs/rules/no-async-describe.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# Disallow async functions passed to describe (no-async-describe)
+
+This rule disallows the use of an async function with `describe`. It usually indicates a copy/paste error or that you're trying to use `describe` for setup code, which should happen in `before` or `beforeEach`. Also, it can lead to [the contained `it` blocks not being picked up](https://github.com/mochajs/mocha/issues/2975).
+
+Example:
+
+```js
+describe(async function () {
```
Comment (lo1tuma) [inline docs/rules/no-async-describe.md]: Is this a valid example? Doesn’t `describe` require a string as the first argument?

File: docs/rules/no-async-describe.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+```
+
+## Rule Details
+
+The rule supports "describe", "context" and "suite" suite function names and different valid suite name prefixes like "skip" or "only".
+
+The following patterns are considered problems, whether or not the function uses `await`:
+
+```js
+describe'something', async function () {
```
Comment (lo1tuma) [inline docs/rules/no-async-describe.md]: Missing `(` after `describe.

File: docs/rules/no-async-describe.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    it('should work', function () {});
+});
+
+describe'something', async () => {
```
Comment (lo1tuma) [inline docs/rules/no-async-describe.md]: Same here.

File: docs/rules/no-async-describe.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    it('should work', function () {});
+});
+```
+
+If the `describe` function does not contain `await`, a fix of removing `async` will be suggested.
+
+This rule looks for every `describe.only`, `it.only`, `suite.only`, `test.only`, `context.only` and `specify.only` occurrences within the source code.
```
Comment (lo1tuma) [inline docs/rules/no-async-describe.md]: Why does it look for `it.only` and `test.only` etc? Maybe the whole paragraph can be removed as the first paragraph in this section already explains that `describe`, `context` and `suite` is supported.

File: docs/rules/no-async-describe.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+Of course there are some edge-cases which can’t be detected by this rule, eg.:
+
+```js
+var describeOnly = describe.only;
+describeOnly.apply(describe);
```
Comment (lo1tuma) [inline docs/rules/no-async-describe.md]: I’m not sure if this example makes sense. It is completely unrelated to `async`.

File: lib/rules/no-async-describe.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    return {
+        CallExpression(node) {
+            const name = astUtils.getNodeName(node.callee);
+
+            if (astUtils.isDescribe(node, additionalSuiteNames(context.settings))) {
+                const fnArg = node.arguments.slice(-1)[0];
+                if (isAsyncFunction(fnArg)) {
+                    context.report({
+                        node: fnArg,
+                        message: `Do not pass an async function to ${name}()`,
```
Comment (lo1tuma) [inline lib/rules/no-async-describe.js]: I would suggest something like `Unexpected async function in ${name}()` to be more in line with messages of other rules.

File: lib/rules/no-async-describe.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            });
+        }
+        return false;
+    }
+
+    function fixAsyncFunction(fixer, fn) {
+        if (!containsDirectAwait(fn.body)) {
+            return fixer.replaceTextRange(
+                [ fn.start, fn.end ],
+                sourceCode.text.slice(fn.range[0], fn.range[1]).replace(/^async /, '')
```
Comment (lo1tuma) [inline lib/rules/no-async-describe.js]: When a function is async, isn’t the first token always the `async` identifier? If so it might be easier to just get the first token of the function node and use `fixer.remove`.

File: lib/rules/no-async-describe.js (hunk 1 above)
Comment (lo1tuma) [reply lib/rules/no-async-describe.js]: Don’t worry, it’s not a requirement to write funky ramda expressions 😉.

File: test/rules/no-async-describe.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
+'use strict';
+
+const RuleTester = require('eslint').RuleTester;
+const rule = require('../../lib/rules/no-async-describe');
+const ruleTester = new RuleTester();
+
+ruleTester.run('no-async-describe', rule, {
+    valid: [
+        'describe()',
+        'describe(function () {})',
```
Comment (lo1tuma) [inline test/rules/no-async-describe.js]: The function should be the second argument of describe.

File: docs/rules/no-async-describe.md (hunk 6 above)
Comment (lo1tuma) [reply docs/rules/no-async-describe.md]: I guess one edge case that we can’t detect might be something like this:

```js
async function mySuite() {
    it('my test', () => {});
}

describe('my suite', mySuite);
```

Comment (papandreou) [reply docs/rules/no-async-describe.md]: D'oh, brain fart. I think I copy pasted from https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/test/rules/no-mocha-arrows.js#L15 and then my brain reset itself into a state where `describe` and `it` don't take a string argument.

Fixed in 7c9b6dc71664d8756fca64ddbbf0bf7657f727e7

File: docs/rules/no-async-describe.md (hunk 2 above)
Comment (papandreou) [reply docs/rules/no-async-describe.md]: No, it isn't :roll_eyes:

Fixed in 7c9b6dc71664d8756fca64ddbbf0bf7657f727e7

File: docs/rules/no-async-describe.md (hunk 5 above)
Comment (papandreou) [reply docs/rules/no-async-describe.md]: Whoops, that part was copy/pasted from the docs for another rule where it did make sense. Maybe I was getting anxious because the other rules are so thoroughly documented :laughing:

File: lib/rules/no-async-describe.js (hunk 7 above)
Comment (papandreou) [reply lib/rules/no-async-describe.js]: Fixed in a5fa46a6fdcf798539c11bfedfdce52eeee4aee9. Looks like I copy/pasted from the `no-mocha-global-arrows` rule.

File: lib/rules/no-async-describe.js (hunk 8 above)
Comment (papandreou) [reply lib/rules/no-async-describe.js]: That's a good point, fixed in 0c113b53b6a7f82efbd3729b04a7f3c59fbb8f3e

We also need go get rid of the whitespace after `async`, so we can remove the range from the start of the `async` token to the start of the next `function` or `(` token.

File: test/rules/no-async-describe.js (hunk 9 above)
Comment (papandreou) [reply test/rules/no-async-describe.js]: Of course! :+1:

Fixed in 7c9b6dc71664d8756fca64ddbbf0bf7657f727e7

**Comment by papandreou** [comment]
Thanks for the feedback, I think I've addressed all of it now.

By the way, the linting setup for this project was a pleasure to work with. I could really feel the presence of the maintainer's opinions in my editor, and that helped a great deal!

The only *slightly* annoying thing is that I can't even run the test suite if everything doesn't lint, because of https://github.com/lo1tuma/eslint-plugin-mocha/blob/27fa81c76d6274d308c4e6110ff99e73895b7409/package.json#L16 -- that gets in the way when you temporarily have some dead code, or if you've commented out some code without the appropriate amount of whitespace after `//` :)

Could it be changed to `posttest` instead?

**Comment by lo1tuma** [comment]
> The only _slightly_ annoying thing is that I can't even run the test suite if everything doesn't lint

That’s true, I usually use only `npm run test:unit` during development, which should even work in watch mode via `npm run test:unit -- --watch`. This command also doesn’t instrument the code for coverage, so it should be much faster.
